### PR TITLE
Added DieLabeller

### PR DESCRIPTION
**Pull request by federico-paolillo**
The DieLabeller class converts an integer result return from an IDie to an arbitrary label of any type.  
It requires the caller to provide a dictionary where the keys are ints and the values are of type TLabel.  
Each key-value pair represents which label will be returned on a rolled value.

The DieLabeller behaves as following:  

- If the value rolled has a corresponding label specified the label is returned.  
- If no label was configure DieLabeller returns default(TLabel).  
- If DieLabeller was configured to throw exception an exception is thrown on any missing value.  

The Moq NuGet package was added to stub an IDie implementation in order to remove non deterministic behavior for the tests and to ensure that only required values are returned from the IDie.

### CODE REVIEW

**Comment by MrTarantula** [comment]
Thanks for the contribution. While I like the concept, there are a few issues. If you create a `LabelledDie` from a `TappersDie`, `TappersDie.Tap()` is no longer exposed. 

I also don't love having to pass a `Dictionary` to the labeller. It can still be included, but I would also like a constructor that can be used like `var labelled = new LabeledDie(die, "these", "are", "some" "labels");`. Nulls could be used for blank spots, and if there are not enough labels, the rest are null.

My dream implementation of this would be to have the labels added like this:

```C#
var die = new GamblersDie(6).WithLabels<string>("these", "are", "my", "labels", null, null);
```

I haven't had much time to think about how that would be implemented.

**Comment by MrTarantula** [comment]
One other nitpick: American English uses one L in labeled, labeler, etc. British, Canadian, and Australian English use two L's. Being an American, I prefer the American version (I'm not trying to sound self-ceneterd but it kind of feels that way, sorry).

**Comment by federico-paolillo** [comment]
Thank you for your feedback, I'll try and see what I can come up with following your suggestions.

**Comment by MrTarantula** [comment]
I've thought about creating some kind of DieBuilder where you could build a die like:

```C#
new Die(20).WithWeights(...).Tappable().Gambler().WithLabels(...).UsingRandomizer(...);
```

The current architecture may not support this very easily.

**Comment by federico-paolillo** [comment]
Well a refactor of that kind is way out of scope of simple pull request, IMHO.  

Following you example of the `WithLabels<string>(...)` method maybe a suitable approach to implement that could be done with extensions methods.  
The `WithLabels<TLabel>(params TLabel[] labels)` extension method could return an `ILabeledDie` or an `ILabeledTappersDie`.  
The `ILabeledTappersDie` will expose and delegate the `Tap()` method to the actual TappersDie.  
The extension method will wrap the `IDie` or `TappersDie` using the `DieLabeller` as I already did.

How does that sound to you ?

**Comment by MrTarantula** [comment]
Yes, my last comment is outside of the scope of this PR. Sorry.

As for the labels, I don't want to have to implement a new labeled interface every time a new die is added. I know this is just a small library and it's trivial to do so, but I would like labels to "just work" when a new kind of die is added. So that if someone takes this library and extends `Die` to `OddNumbersWeightedEveryThirdRollDie` they won't need to add `ILabeledOddNumbersWeightedEveryThirdRollDie` interface. 

Does that make sense?

**Comment by MrTarantula** [comment]
Which I understand this PR does work for most of the dice, but I don't want to break behavior of things like `TappersDie` and I also don't want to add a no-op `Tap()` to `Die` just for the sake of the `TappersDie`.

**Comment by MrTarantula** [comment]
NDice is nowhere near a 1.0 release, so this can go in, after a few nits.

File: NDice/DieLabeller.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+using System.Text;
+
+namespace NDice
+{
+	/// <summary>
+	/// Labels a Die roll result to a value of type TLabel
+	/// </summary>
+    public class DieLabeller<TLabel>
+    {
+		private readonly IDie die = null;
```
Comment (MrTarantula) [inline NDice/DieLabeller.cs]: No need for assignments. Declaring is fine. Same for the all of the private variables.

Comment (MrTarantula) [reply NDice/DieLabeller.cs]: Also, the convention in this lib is private vars are underscored: `_die`, `_labels`, and `throwsIfLabelMissing`. Then `this` can be removed in the constructors.

File: NDice/DieLabeller.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+		private readonly bool throwsIfLabelMissing = false;
+
+		public DieLabeller(IDie die, IReadOnlyDictionary<int, TLabel> labels)
+			: this(die, labels, throwsIfLabelMissing: false)
+		{
+		}
+
+		public DieLabeller(IDie die, IReadOnlyDictionary<int, TLabel> labels, bool throwsIfLabelMissing)
+		{
+			this.die = die ?? throw new ArgumentNullException(nameof(die), "An IDie instance must be supplied.");
```
Comment (MrTarantula) [inline NDice/DieLabeller.cs]: Shouldn't be null since it's required.

File: NDice/LabelMissingException.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+		{
+		}
+
+		public LabelMissingException(string message, System.Exception inner) 
+			: base(message, inner)
+		{
+		}
+
+		protected LabelMissingException(
+		  System.Runtime.Serialization.SerializationInfo info,
```
Comment (MrTarantula) [inline NDice/LabelMissingException.cs]: `using System.Runtime.Serialization;` at the top of the file.

**Comment by MrTarantula** [comment]
Please update the readme as well. You'll need to rebase on my master. I just added a few things.

**Comment by federico-paolillo** [comment]
Great, thank you for your feedback ! I'll try to update the PR as soon as possible this week.

**Comment by federico-paolillo** [comment]
Updated branch after PR review

File: NDice/DieLabeler.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace NDice
+{
+	/// <summary>
+	/// Labels a Die roll result to a value of type TLabel
+	/// </summary>
+    public class DieLabeler<TLabel>
+    {
+		private readonly IDie _die = null;
+		private readonly IReadOnlyDictionary<int, TLabel> _labels = null;
```
Comment (MrTarantula) [inline NDice/DieLabeler.cs]: Default values aren't needed for any of these. Also, do we want to make labels public? Wouldn't it be useful to be able to access all labels?

**Comment by MrTarantula** [comment]
Also, `DieLabeler` should be added to the readme.

**Comment by federico-paolillo** [comment]
Regarding `DieLabeler` in the readme file, do you want me to provide a small example or should I add another section as you did for `IRandomizable` ?

**Comment by federico-paolillo** [comment]
Updated branch after PR review, sorry for the delay !

File: README.md [hunk 5]
```diff
@@ -113,7 +142,7 @@ Double Deuce<br>![Double Deuce](double2.png) | <pre>var d2Die = new WeightedDie(
 - [ ] More real world examples
 - [x] Percentages/ratios for weight
 - [ ] Include common dice like the examples (another package)
-- [ ] Built-in labels
+- [x] Built-in labels
```
Comment (federico-paolillo) [inline README.md]: I took the liberty to mark this feature as complete, although I'm not really sure if I really added built-in labels the way you would like them.

**Comment by MrTarantula** [comment]
Looks good. Thanks.

**Comment by MrTarantula** [comment]
[v0.3.0](https://www.nuget.org/packages/NDice/) has been pushed to NuGet with `DieLabeler`.

**Pull request merged**